napalm_eos/eos.py

```diff
@@ -1201,3 +1201,162 @@ def traceroute(self, destination, source='', ttl=0, timeout=0):
                 previous_probe_ip_address = ip_address
 
         return traceroute_result
+
+    def get_bgp_neighbors_detail(self, neighbor_address=''):
+        """Function to return detailed BGP information.
+
+        Information returned by this function can be about a single peer or
+        about all peers. This can be controlled using the following:
+
+        :params neighbor_address: Retuns the statistics for a specific
+                                  BGP neighbor or for all BGP neighbors.
+        """
+
+        peer_ver = None
+        is_all = True if not neighbor_address else False
+
+        if is_all:
+            command = ['show ip bgp neighbors vrf all']
+        else:
+            peer_ver = IPAddress(neighbor_address).version
+            if peer_ver == 4:
+                command = ['show ip bgp neighbors %s vrf all' %
+                           neighbor_address]
+            else:
+                command = ['show ipv6 bgp neighbors %s vrf all' %
+                           neighbor_address]
+
+        raw_output = (
+            self.device.run_commands(command, encoding='text')[0]['output'])
+
+        bgp_detail_info = {}
+        int_fields = ['accepted_prefix_count', 'local_as', 'remote_as',
+                      'local_port', 'remote_port', 'local_port',
+                      'input_messages', 'output_messages', 'input_updates',
+                      'output_updates', 'messages_queued_out', 'holdtime',
+                      'configured_holdtime', 'keepalive',
+                      'configured_keepalive', 'advertise_prefix_count',
+                      'received_prefix_count']
+
+        # Parsing through the raw bgp output
+        if peer_ver == 6:
+            neighbor_match_string = (
+                r'BGP neighbor is (.*), remote AS (\d+),')
+        else:
+            neighbor_match_string = (
+                r'BGP neighbor is (\d+.\d+.\d+.\d+), remote AS (\d+),')
+
+        parsed_peers = re.findall(neighbor_match_string, raw_output)
+
+        # Adding remote ASN's as top level keys in dict
+        for item in parsed_peers:
+            if item[1] not in bgp_detail_info:
+                bgp_detail_info[int(item[1])] = []
+
+        if is_all:
+            # Issuing commands per peer
+            for item in parsed_peers:
+                if peer_ver == 6:
+                    command = ['show ipv6 bgp neighbors %s vrf all' % item[0]]
```

Comment:
Yes, my thoughts are the same: there's absolutely no difference in terms of information per neighbor between ```sh ip bgp neighbors vrf all``` and ```sh ip bgp neighbors %s vrf all```.
When retrieving per neighbor, not only that you stress the CPU of the router (and you should definitely avoid that, it has to be busy with other processes not serving unnecessary request), but also a ping-pong between the router and your server leads to an extremely slow function. You may not feel that if you test on a local machine, or on a remove machine with a tiny number of peers.
I run against a router physically placed geographically a couple of thousands of km away from the server, having a pretty modest number of neighbors:

```python
>>> start = time.time(); d = e.get_bgp_neighbors_detail(); time.time() - start
90.02689695358276
```

That's huge!

On the other hand, the servers are usually powerful enough and parsing a large text file in the worst possible case would take 2-3 seconds.

So: please retrieve all text output at once and use it to extract data.